Quick and dirty script to dump IDs

kaapstorm opened the pull request:
##### SUMMARY
A quick and dirty script to dump IDs for forms, visits, and child cases as CSV

(For reference. Not for merging.)

kaapstorm committed:
quick and dirty script to dump IDs

esoergel commented on custom/rec/management/commands/export_visit_child.py:
```diff
@@ -0,0 +1,69 @@
+    if "case_case_child" in couch_form.form_json["form"]:
+        return couch_form.form_json["form"]["case_case_child"]["case"]["@case_id"]
+    imci_visit_id = get_imci_visit_id(couch_form)
+    imci_visit = CommCareCase.get(imci_visit_id)
```
This might make it really slow assuming there are a large number of forms at play.  Would it be straightforward to do a bulk lookup for each form chunk?

kaapstorm replied:
Good idea. [d7ea6a0](https://github.com/dimagi/commcare-hq/pull/25617/commits/d7ea6a0c514722578ad30c78d07fd15aad04369e) fetches all the visit cases we need in one go for each chunk of forms. (It only fetches those cases for which we don't already have rec_child_id.)

kaapstorm committed:
Get visit cases in chunks

kaapstorm commented:
(This script fetches forms and cases from Couch. Once the project has been migrated to SQL, we'd want to use SQL dbaccessors and `CommCareCaseSQL`)

kaapstorm committed:
Merge branch 'master' into nh/jessica_csv

kaapstorm committed:
Add all XMLNSes

kaapstorm committed:
If you can OR sets, why can't you OR dictionaries?

kaapstorm committed:
Organize forms

emord commented on custom/rec/collections.py:
```diff
@@ -0,0 +1,17 @@
+        return dicts[0]
+    d1 = dicts[0]
+    dn = dicts_or(*dicts[1:])
+    return {k: d1[k] if k in d1 else dn[k] for k in d1.keys() | dn.keys()}
```
Is this any different than:

```python
ret = {}
for d in reverse(dicts):
    ret.update(d)
return ret
```

I reversed ret because `update` appears to have the opposite behavior of what you were going for.

However looking at your only usage of this in the next commit, I don't think that order matters

esoergel replied:
this looks fun - I'll throw my hat in the ring with
```python
 def dicts_or(*dicts): 
    return dict(chain(*(d.items() for d in dicts)))
```
or the slightly less fancy
```python
 def dicts_or(*dicts): 
    return {k: v for d in dicts for k, v in d.items()} 
```
possibly also with a `reversed` in there if necessary as @emord described

kaapstorm replied:
Oh yeah! Those are much better! Nice!

kaapstorm committed:
:black_heart: short dict func

This reverses `dicts` even though we don't really need to for our
usecase, because otherwise it wouldn't be an "or". It would be like an
"update, reduced", which is 6 characters shorter, but just doesn't
sound as cool.
